### ISSUE TITLE
Assertion failure on glob assignment

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9515]
Migrated from [rt.perl.org#59718](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=59718) (status was 'resolved')

Searchable as RT59718$

**Comment by p5pRT** [comment]
### From mmaslano@redhat.com

#### Created by mmaslano@redhat.com



Description From  Michael Chapman \(redhat\-bugzilla@&#8203;very\.puzzling\.org\) 
2008\-10\-06 23&#8203;:26&#8203;:56 EDT

Description of problem&#8203;:

The following Perl script generates an assertion failure&#8203;:

$ perl \-e 'BEGIN \{ $&#8203;::\{A\} = \\1 \}; index ""\, A; \*B = $&#8203;::\{A\}'
Assertion \(\(\(\(shplep\)\->sv\_flags & \(0x00004000|0x00008000\)\) == 0x00008000\) &&
\(\(\(svtype\)\(\(shplep\)\->sv\_flags & 0xff\)\) == SVt\_PVGV ||
\(\(svtype\)\(\(shplep\)\->sv\_flags & 0xff\)\) == SVt\_PVLV\)\) failed&#8203;: file "sv\.c"\, 
line
3212 at \-e line 1\.

Removing the index statement clears the error&#8203;:

$ perl \-e 'BEGIN \{ $&#8203;::\{A\} = \\1 \}; \*B = $&#8203;::\{A\}'
\# no error

Version\-Release number of selected component \(if applicable\)&#8203;:

$ rpm \-q perl
perl\-5\.10\.0\-34\.fc9\.i386


Additional info&#8203;:

The BEGIN block is effectively equivalent to&#8203;:

&nbsp;   use constant A => 1;

This also generates the same error\.

index does not need to be executed\, it merely has to be used on A 
somewhere in
the script\. For instance&#8203;:

&nbsp;   index ""\, A if 0;

produces the same error\. The problem appears to be related to Perl's 
upgrading
of the constant to attach Boyer\-Moore magic&#8203;:

$ perl \-mDevel&#8203;::Peek=Dump \-e 'BEGIN \{ $&#8203;::\{A\} = \\1 \} Dump $&#8203;::\{A\}'  \# good
SV = RV\(0x9638200\) at 0x96381f4
&nbsp;   REFCNT = 1
&nbsp;   FLAGS = \(ROK\)
&nbsp;   RV = 0x9638234
&nbsp;   SV = IV\(0x9638230\) at 0x9638234
&nbsp;     REFCNT = 1
&nbsp;     FLAGS = \(IOK\,READONLY\,pIOK\)
&nbsp;     IV = 1
$ perl \-mDevel&#8203;::Peek=Dump \-e 'BEGIN \{ $&#8203;::\{A\} = \\1 \} index ""\, A if 0; Dump
$&#8203;::\{A\}'  \# bad
SV = RV\(0x80a1208\) at 0x80a11fc
&nbsp;   REFCNT = 1
&nbsp;   FLAGS = \(ROK\)
&nbsp;   RV = 0x80a123c
&nbsp;   SV = PVGV\(0x8099720\) at 0x80a123c
&nbsp;     REFCNT = 2
&nbsp;     FLAGS = \(PADTMP\,SMG\,POK\,READONLY\,pPOK\,VALID\,EVALED\)
&nbsp;     IV = 100
&nbsp;     PV = 0x80a56ac "1"\\0
&nbsp;     CUR = 1
&nbsp;     LEN = 4
&nbsp;     MAGIC = 0x80a9fb4
&nbsp;       MG\_VIRTUAL = &PL\_vtbl\_bm
&nbsp;       MG\_TYPE = PERL\_MAGIC\_bm\(B\)
&nbsp;     FLAGS = 0
&nbsp;     RARE = 49
&nbsp;     PREVIOUS = 0
&nbsp;     USEFUL = 100

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=low

This perlbug was built using Perl 5.10.0 in the Fedora build system.
It is being executed now by Perl 5.10.0 - Wed Sep 17 10:32:41 EDT 2008.

Site configuration information for perl 5.10.0:

Configured by Red Hat, Inc. at Wed Sep 17 10:32:41 EDT 2008.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
   Platform:
     osname=linux, osvers=2.6.18-92.1.10.el5, 
archname=i386-linux-thread-multi
     uname='linux x86-6.fedora.phx.redhat.com 2.6.18-92.1.10.el5 #1 smp 
wed jul 23 03:56:11 edt 2008 i686 i686 i386 gnulinux '
     config_args='-des -Doptimize=-O2 -g -pipe -Wall 
-Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector 
--param=ssp-buffer-size=4 -m32 -march=i386 -mtune=generic 
-fasynchronous-unwind-tables -DPERL_USE_SAFE_PUTENV -Dversion=5.10.0 
-Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red 
Hat, Inc. -Dinstallprefix=/usr -Dprefix=/usr 
-Dprivlib=/usr/lib/perl5/5.10.0 
-Dsitelib=/usr/local/lib/perl5/site_perl/5.10.0 
-Dvendorlib=/usr/lib/perl5/vendor_perl/5.10.0 
-Darchlib=/usr/lib/perl5/5.10.0/i386-linux-thread-multi 
-Dsitearch=/usr/local/lib/perl5/site_perl/5.10.0/i386-linux-thread-multi 
-Dvendorarch=/usr/lib/perl5/vendor_perl/5.10.0/i386-linux-thread-multi 
-Darchname=i386-linux-thread-multi 
-Dotherlibdirs=/usr/local/lib/perl5/site_perl:/usr/lib/perl5/site_perl 
-Dvendorprefix=/usr -Dsiteprefix=/usr/local -Duseshrplib -Dusethreads 
-Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db 
-Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio 
-Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly 
-Dpager=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_endhostent_r_proto 
-Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto 
-Ud_endservent_r_proto -Ud_setservent_r_proto -Dscriptdir=/usr/bin'
     hint=recommended, useposix=true, d_sigaction=define
     useithreads=define, usemultiplicity=define
     useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
     use64bitint=undef, use64bitall=undef, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING 
-fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE 
-D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
     optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions 
-fstack-protector --param=ssp-buffer-size=4 -m32 -march=i386 
-mtune=generic -fasynchronous-unwind-tables -DPERL_USE_SAFE_PUTENV',
     cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING 
-fno-strict-aliasing -pipe -I/usr/local/include -I/usr/include/gdbm'
     ccversion='', gccversion='4.3.0 20080428 (Red Hat 4.3.0-8)', 
gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=4, prototype=define
   Linker and Libraries:
     ld='gcc', ldflags =' -L/usr/local/lib'
     libpth=/usr/local/lib /lib /usr/lib
     libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
     perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
     libc=/lib/libc-2.8.so, so=so, useshrplib=true, libperl=libperl.so
     gnulibc_version='2.8'
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E 
-Wl,-rpath,/usr/lib/perl5/5.10.0/i386-linux-thread-multi/CORE'
     cccdlflags='-fPIC', lddlflags='-shared -O2 -g -pipe -Wall 
-Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector 
--param=ssp-buffer-size=4 -m32 -march=i386 -mtune=generic 
-fasynchronous-unwind-tables -DPERL_USE_SAFE_PUTENV -L/usr/local/lib'

Locally applied patches:



@INC for perl 5.10.0:
     /usr/lib/perl5/5.10.0/i386-linux-thread-multi
     /usr/lib/perl5/5.10.0
     /usr/local/lib/perl5/site_perl/5.10.0/i386-linux-thread-multi
     /usr/local/lib/perl5/site_perl/5.10.0
     /usr/lib/perl5/vendor_perl/5.10.0/i386-linux-thread-multi
     /usr/lib/perl5/vendor_perl/5.10.0
     /usr/lib/perl5/vendor_perl
     /usr/local/lib/perl5/site_perl/5.10.0/i386-linux-thread-multi
     /usr/local/lib/perl5/site_perl/5.10.0
     /usr/local/lib/perl5/site_perl
     /usr/lib/perl5/site_perl/5.10.0/i386-linux-thread-multi
     /usr/lib/perl5/site_perl/5.10.0
     /usr/lib/perl5/site_perl
     .


Environment for perl 5.10.0:
     HOME=/home/marca
     LANG=en_US.UTF-8
     LANGUAGE=en_US
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
 
PATH=/usr/lib/qt4/bin:/usr/lib/qt-3.3/bin:/usr/kerberos/bin:/usr/local/bin:/usr/bin:/bin:/home/marca/bin
     PERL_BADLANG (unset)
     SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From ben@morrow.me.uk


Quoth perl5\-porters@&#8203;perl\.org&#8203;:
> 
> The following Perl script generates an assertion failure&#8203;:
> 
> $ perl \-e 'BEGIN \{ $&#8203;::\{A\} = \\1 \}; index ""\, A; \*B = $&#8203;::\{A\}'
> Assertion \(\(\(\(shplep\)\->sv\_flags & \(0x00004000|0x00008000\)\) == 0x00008000\) &&
> \(\(\(svtype\)\(\(shplep\)\->sv\_flags & 0xff\)\) == SVt\_PVGV ||
> \(\(svtype\)\(\(shplep\)\->sv\_flags & 0xff\)\) == SVt\_PVLV\)\) failed&#8203;: file "sv\.c"\, 
> line
> 3212 at \-e line 1\.
> 
> Removing the index statement clears the error&#8203;:
> 
> $ perl \-e 'BEGIN \{ $&#8203;::\{A\} = \\1 \}; \*B = $&#8203;::\{A\}'
> \# no error

I believe this was fixed with \#34213\, which went into 5\.10 as \#34295\.

Ben

\-\- 
The Earth is degenerating these days\. Bribery and corruption abound\.
Children no longer mind their parents\, every man wants to write a book\,
and it is evident that the end of the world is fast approaching\.
&nbsp;       Assyrian stone tablet\, c\.2800 BC                        ben@&#8203;morrow\.me\.uk

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From mmaslano@redhat.com

Ben Morrow via RT wrote&#8203;:
> Quoth perl5\-porters@&#8203;perl\.org&#8203;:
>> The following Perl script generates an assertion failure&#8203;:
>>
>> $ perl \-e 'BEGIN \{ $&#8203;::\{A\} = \\1 \}; index ""\, A; \*B = $&#8203;::\{A\}'
>> Assertion \(\(\(\(shplep\)\->sv\_flags & \(0x00004000|0x00008000\)\) == 0x00008000\) &&
>> \(\(\(svtype\)\(\(shplep\)\->sv\_flags & 0xff\)\) == SVt\_PVGV ||
>> \(\(svtype\)\(\(shplep\)\->sv\_flags & 0xff\)\) == SVt\_PVLV\)\) failed&#8203;: file "sv\.c"\, 
>> line
>> 3212 at \-e line 1\.
>>
>> Removing the index statement clears the error&#8203;:
>>
>> $ perl \-e 'BEGIN \{ $&#8203;::\{A\} = \\1 \}; \*B = $&#8203;::\{A\}'
>> \# no error
> 
> I believe this was fixed with \#34213\, which went into 5\.10 as \#34295\.
> 
> Ben
> 
True\. Thanks for pointing this change\.

\-\- 
Marcela Mašláňová
BaseOS team Brno

**Comment by p5pRT** [comment]
@chipdude - Status changed from 'open' to 'resolved'
